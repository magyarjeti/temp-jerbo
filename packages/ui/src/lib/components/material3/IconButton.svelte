<script lang="ts">
	import { concatClasses } from '@packages/toolbox/concatClasses';
	import JetiIcon from '../JetiIcon.svelte';
	import StateLayer from './StateLayer.svelte';
	import type { HTMLButtonAttributes, HTMLAnchorAttributes } from 'svelte/elements';
	import type { Snippet } from 'svelte';
	import { isAnchorAttrs, isButtonAttrs } from '../../utils/components';

	type Props = {
		variant?: 'standard' | 'outlined' | 'filled';
		disabled?: boolean | undefined;
		icon?: string | undefined | null;
		selected?: boolean;
		children?: Snippet;
	} & (HTMLButtonAttributes | HTMLAnchorAttributes);

	let {
		variant = 'standard',
		disabled,
		icon,
		selected = false,
		children,
		...attrs
	}: Props = $props();

	const variantClassMap: Record<typeof variant, string> = $derived({
		standard: concatClasses('text-primary', selected && 'bg-on-surface-variant/12'),
		outlined: concatClasses(
			'border border-outline disabled:border-on-surface/12',
			selected && 'bg-on-surface-variant/12',
		),
		filled: concatClasses(
			'text-primary bg-surface-container-highest disabled:border-on-surface/12',
			selected && 'bg-primary text-on-primary',
		),
	});
	const stateLayerClasses: Record<typeof variant, string> = {
		standard: 'bg-on-surface-variant',
		outlined: 'bg-on-surface-variant',
		filled: 'bg-primary',
	};
	const baseClasses =
		'relative shape-full typo-label-medium h-10 w-10 text-[1em] disabled:text-on-surface/38 inline-flex max-w-full cursor-pointer items-center justify-center';
	const classes = $derived(concatClasses(attrs.class, baseClasses, variantClassMap[variant]));
</script>

{#snippet inside()}
	<StateLayer class={concatClasses(stateLayerClasses[variant], 'shape-full')} />
	{#if children}
		{@render children()}
	{:else}
		<JetiIcon>{icon}</JetiIcon>
	{/if}
{/snippet}

{#if isAnchorAttrs(attrs)}
	<a
		aria-disabled={disabled}
		{...attrs}
		class={concatClasses(
			classes,
			disabled && 'text-on-surface/38',
			disabled && variant !== 'standard' && 'border-on-surface/12',
		)}
	>
		{@render inside()}
	</a>
{:else if isButtonAttrs(attrs)}
	<button
		class={classes}
		aria-disabled={disabled}
		{disabled}
		type={attrs.type || 'button'}
		{...attrs}
	>
		{@render inside()}
	</button>
{/if}
