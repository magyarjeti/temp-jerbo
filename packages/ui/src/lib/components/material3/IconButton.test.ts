import { render } from '@testing-library/svelte';
import { expect, test } from 'vitest';
import IconButton from './IconButton.svelte';

test('Standard IconButton snapshot', () => {
	const { container } = render(IconButton, { props: { variant: 'standard' } });
	expect(container).toMatchSnapshot();
});

test('Outlined IconButton snapshot', () => {
	const { container } = render(IconButton, { props: { variant: 'outlined' } });
	expect(container).toMatchSnapshot();
});

test('Filled IconButton snapshot', () => {
	const { container } = render(IconButton, { props: { variant: 'filled' } });
	expect(container).toMatchSnapshot();
});
