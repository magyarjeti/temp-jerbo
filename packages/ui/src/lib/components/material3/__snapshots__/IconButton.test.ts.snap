// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Filled IconButton snapshot 1`] = `
<div>
  <button
    class="cursor-pointer inline-flex items-center justify-center shape-full typo-label-medium h-10 w-10 text-[1em] bg-primary text-on-primary hover:text-on-primary state-layers state-layer-hover-on-primary state-layer-focus-on-primary disabled:text-on-surface/38 disabled:border-on-surface/12"
  >
    <span
      class="jeti-icon svelte-1pzkbd4"
    />
  </button>
  
</div>
`;

exports[`Outlined IconButton snapshot 1`] = `
<div>
  <button
    class="cursor-pointer inline-flex items-center justify-center shape-full typo-label-medium h-10 w-10 text-[1em] border border-outline state-layers state-layer-hover-on-surface-variant state-layer-focus-on-surface-variant disabled:text-on-surface/38 disabled:border-on-surface/12"
  >
    <span
      class="jeti-icon svelte-1pzkbd4"
    />
  </button>
  
</div>
`;

exports[`Standard IconButton snapshot 1`] = `
<div>
  <button
    class="cursor-pointer inline-flex items-center justify-center shape-full typo-label-medium h-10 w-10 text-[1em] text-primary state-layers state-layer-hover-on-surface-variant state-layer-focus-on-surface-variant disabled:text-on-surface/38"
  >
    <span
      class="jeti-icon svelte-1pzkbd4"
    />
  </button>
  
</div>
`;
